contrib/masterfiles-stage/common.sh

```diff
@@ -12,146 +12,244 @@ if [ ! "$debug_mode" == "true" ]; then
-	  # Not sure we would want to rebase changes either, we want a clean sync with the upstream
-          # The above line appears to be a mistake as it will ALWAYS fail if given a tagname or commit hash.
-          # It will succeed if given "tags/tagname" or if given a branch name, but never with a bare tagname or commit hash.
-          git reset -q --hard "${GIT_TAG_OR_COMMIT}" || error_exit "Failed: git reset -q --hard ${GIT_TAG_OR_COMMIT}"
-          git clean -f || error_exit "Failed: git clean -f"
-          git clean -fd || error_exit "Failed: git clean -fd"
-      ) || error_exit "Failed to stage '${GIT_TAG_OR_COMMIT}' in '${STAGING_DIR}' from '${GIT_URL}'" 
-
-  # If we don't have a git clone wipe the directory and create a new clone and
-  # ensure the proper branch/tag is checked out.
+  if [ -d "${MASTERDIR}" ] ; then
+    # Put tmpdir in MASTERDIR's parent dir to avoid crossing filesystem boundaries
+    # (which could be a danger if we used /tmp).
+    third_dir="$(mktemp -d --tmpdir="$(dirname "${MASTERDIR}")" )"
+
+    mv "${MASTERDIR}" "${third_dir}/momentary"  || error_exit "Can't mv ${MASTERDIR} to ${third_dir}"
+    mv "${STAGING_DIR}" "${MASTERDIR}"          || error_exit "Can't mv ${STAGING_DIR} to ${MASTERDIR}"
+    mv "${third_dir}/momentary" "${STAGING_DIR}"   # We don't care if this fails;
+    rm -rf "${third_dir}"                          # we're going to remove third_dir anyways.
   else
-      rm -rf "${STAGING_DIR}"/* "${STAGING_DIR}"/.??*
-      git clone --no-hardlinks "${GIT_URL}" "${STAGING_DIR}" && cd "${STAGING_DIR}" && git checkout "${GIT_TAG_OR_COMMIT}" || error_exit "Failed to stage '${GIT_TAG_OR_COMMIT}' from '${GIT_URL}'"
+    mv "${STAGING_DIR}" "${MASTERDIR}"          || error_exit "Can't mv ${STAGING_DIR} to ${MASTERDIR}"
   fi
+}
+
+######################################################
+##           VCS_TYPE-based main functions           #
+######################################################
+
+git_stage_policy_channels() {
+  # Created by Mike Weilgart
+  #
+  # This "VCS_TYPE-based" function is called from masterfiles-stage.sh.
+  #
+  # This function stages multiple policy channels each to its masterdir,
+  # based on the values specified in the bash array "channel_config"
+  # defined in the params file, and the GIT_URL value also set in that file.
+  #
+  # (See the example git policy channels params file.)
+  #
+  # The value of MASTERDIR that is assigned in masterfiles-stage.sh
+  # is ****IGNORED**** by this function, since there is a separate
+  # MASTERDIR for each separate policy channel.
+  #
+  # Example value for channel_config:
+  # (This is a single value, split into three lines for readability.)
+  #
+  # channel_config=()
+  # channel_config+=( "/var/cfengine/masterfiles" "my_git_tag" )
+  # channel_config+=( "/var/cfengine/policy_channels/channel_1" "my_git_branch" )
```

Comment:
@tzz, I've implemented the bash array change.  Also, @nickanderson, this should fix the issue with Design Center (since I skip using `$ROOT` entirely) and also allow for the `cf_promises_release_id` to be populated correctly.

I only changed `git_stage_policy_channels`, not the `GIT` or `GIT_TAG_OR_COMMIT` vcs types.